lib/ddtrace/method_tracer.rb

```diff
@@ -0,0 +1,123 @@
+# frozen_string_literal: true
+
+module Datadog
+  # \MethodTracer adds helpers for unobtrusive methods tracking.
+  #
+  # == Example
+  #
+  #   Datadog::MethodTracer.trace_methods(ExampleClass, :some_instance_method, :other_instance_method)
+  #   Datadog::MethodTracer.trace_singleton_methods(ExampleClass, :some_class_method, :other_class_method)
+  #
+  module MethodTracer
+    module_function
+
+    def trace_methods(klass, *method_names)
+      instrumentation.trace_methods klass, *method_names
+    end
+
+    def trace_singleton_methods(klass, *method_names)
+      instrumentation.trace_singleton_methods klass, *method_names
+    end
+
+    def instrumentation
+      @_instrumentation ||= Instrumentation.new
+    end
+
+    # \Mixin adds helpers for unobtrusive methods tracking.
+    #
+    # == Example
+    #
+    #   class Foo
+    #     include Datadog::MethodTracer
+    #     trace_methods(:some_instance_method, :other_instance_method)
+    #     trace_singleton_methods(:some_class_method, :other_class_method)
+    #
+    module Mixin
+      def self.included(base)
+        base.extend(ClassMethods)
+      end
+
+      # Added to extended class
+      module ClassMethods
+        def trace_methods(*method_names)
+          MethodTracer.instrumentation.trace_methods self, *method_names
+        end
+
+        def trace_singleton_methods(*method_names)
+          MethodTracer.instrumentation.trace_singleton_methods self, *method_names
+        end
+      end
+    end
+
+    # \Instrumentation for method tracing.
+    class Instrumentation
+      OPERATION_NAME = 'method.call'.freeze
+      private_constant :OPERATION_NAME
+
+      def trace_methods(klass, *method_names)
+        method_names.each { |method_name| instrument_method(klass, method_name) }
+        prepend_container(klass)
+      end
+
+      def trace_singleton_methods(klass, *method_names)
+        method_names.each { |method_name| instrument_singleton_method(klass, method_name) }
+        prepend_container(klass)
+      end
+
+      private
+
+      def instrument_method(klass, method)
+        Container.create_method(method) do |*args, &block|
```

Comment:
Wouldn't `Container` here always refer to the same `Datadog::MethodTracer::Container` module? Meaning `create_method` would be creating a method on a globally shared singleton `Container`?

If this assumption of mine is correct, and `Instrumentation#trace_methods` and `Instrumentation#trace_singleton_methods` have to create new modules for each `klass.send(:prepend, Container)` call, I think that we should ensure that `#trace_methods` and `#trace_singleton_methods` reuse the prepended module if it's already there.

Reply:
Your assumption is correct [added condition](https://github.com/DataDog/dd-trace-rb/pull/1215/files#diff-ccf38a2ec0604d3bdf40d8ed352af6fca58d18fcf29578f82c5b0c5155b4a864R70)